### PR TITLE
molecule: init at 5.1.0, python310Packages.molecule-plugins: init at 23.4.1

### PR DESCRIPTION
**Pull request by dawidd6**
###### Description of changes

This PR is a successor to https://github.com/NixOS/nixpkgs/pull/156612.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dawidd6** [comment]
@ofborg build molecule

**Comment by Et7f3** [comment]
I don't think ofborg understand this commit syntax. Can you split in two commit or use pkg1,pkg2: init at v1,v2 (ofborg understand this).

i will test molecule soon. Can we test molecules plugin without molecules ? I know how to test molecule plugin but wanted to test them separately.

Why are you reversing the deps ?

**Comment by dawidd6** [comment]
`molecule-plugins` depends on `molecule` in upstream repo. I'm not sure if plugins can be tested separately.

Maybe I'm not knowledgeable enough but it seemed easier to implement reversing the dependencies in nixpkgs. Actually I don't know how it would work (if it would work at all) when someone wanted to get molecule with plugins installed. I know one can install python with some packages (like requests) but I have no idea if it is possible to do something like: `nix profile install nixpkgs#molecule nixpkgs#python3.pkgs.molecule-plugins`. Would that get one molecule with plugins? Maybe my understanding is limited.

**Comment by Et7f3** [comment]
What can be done is naming the top level attribute molecule and have it reference molecule-plugins so when someone install molecule it get a working molecule. Something like we have with cc-unwrapped

**Comment by dawidd6** [comment]
Neither `python3Packages.molecule,python3Packages.molecule-plugins: init at 5.1.0,23.4.1` nor `python3Packages.molecule: init at 5.1.0, python3Packages.molecule-plugins: init at 23.4.1` worked for ofborg to find changed paths :thinking:

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -6455,6 +6455,10 @@ self: super: with self; {
 
   mohawk = callPackage ../development/python-modules/mohawk { };
 
+  molecule = callPackage ../development/python-modules/molecule { };
```
Comment (SuperSandro2000) [inline pkgs/top-level/python-packages.nix:6458]: This is also used as a library, not only as an application?

Comment (dawidd6) [reply pkgs/top-level/python-packages.nix:6458]: Well, upstream lists this:

image

So I would guess yes.

Comment (Et7f3) [reply pkgs/top-level/python-packages.nix:6458]: Molecule plugins use it as library

**Review by SuperSandro2000 — COMMENTED** [review]
each new package should be it's own commit then also ofborg works like expected

**Comment by Et7f3** [comment]
```
molecule reset
Traceback (most recent call last):
  File "/nix/store/qrdxj46nz3s5f7wgzl6n4y124iy47lvk-python3.10-molecule-5.1.0/bin/.molecule-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/f0sfxf4yjc0xsylxxn0syrpmkb5wlq01-python3.10-click-8.1.3/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/nix/store/f0sfxf4yjc0xsylxxn0syrpmkb5wlq01-python3.10-click-8.1.3/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/nix/store/f0sfxf4yjc0xsylxxn0syrpmkb5wlq01-python3.10-click-8.1.3/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/nix/store/f0sfxf4yjc0xsylxxn0syrpmkb5wlq01-python3.10-click-8.1.3/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/nix/store/f0sfxf4yjc0xsylxxn0syrpmkb5wlq01-python3.10-click-8.1.3/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/nix/store/f0sfxf4yjc0xsylxxn0syrpmkb5wlq01-python3.10-click-8.1.3/lib/python3.10/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/nix/store/qrdxj46nz3s5f7wgzl6n4y124iy47lvk-python3.10-molecule-5.1.0/lib/python3.10/site-packages/molecule/command/reset.py", line 48, in reset
    driver.reset()
  File "/nix/store/wy6k5cyyc1iqrgggpqx8ibkxrl3iv420-python3.10-molecule-plugins-23.4.1/lib/python3.10/site-packages/molecule_plugins/docker/driver.py", line 256, in reset
    import docker
ModuleNotFoundError: No module named 'docker'
```

**Review by natsukium — CHANGES_REQUESTED** [review]
Would you please split the commit according to the [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) and [SuperSandro2000's comment](https://github.com/NixOS/nixpkgs/pull/241313#pullrequestreview-1530885739)?
Thanks.

File: pkgs/development/python-modules/molecule/plugins.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  # reverse the dependency
+  pythonRemoveDeps = [
+    "molecule"
+  ];
+
+  nativeBuildInputs = [
+    pythonRelaxDepsHook
+    setuptools-scm
+  ];
+
```
Comment (natsukium) [inline pkgs/development/python-modules/molecule/plugins.nix:27]: Can we set `passthru.optional-dependencies`?
https://github.com/ansible-community/molecule-plugins/blob/7eab71d0da547914e46a1110c56d4ac949b9bce9/pyproject.toml#L51

Comment (dawidd6) [reply pkgs/development/python-modules/molecule/plugins.nix:29]: https://github.com/NixOS/nixpkgs/pull/241313/commits/1d37534b1c1ac6259ee189a6c3971763765a1fc4

**Review by natsukium — APPROVED** [review]
The packaging seems fine, so I'll merge it.
Please open a PR if there is further progress on https://github.com/NixOS/nixpkgs/pull/241313#issuecomment-1636735458.

**Pull request merged**